Migrate from nosetest to pytest

aucampia opened the pull request:
This patch replace all uses of nose with pytest. It also includes a
pytest plugin for creating EARL reports for tests with a `rdf_test_uri`
parameter.

Some caveats:

- HTML report directory is now htmlcov instead of coverage
- ~~There is some warning related to the EARL reporting plugin which I can't quite figure out:~~
  ```
  .venv/lib64/python3.7/site-packages/_pytest/config/__init__.py:676
    /home/iwana/sw/d/github.com/iafork/rdflib/.venv/lib64/python3.7/site-packages/_pytest/config/__init__.py:676: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: test.earl
      self.import_plugin(import_spec)
  ```
  ~~This is not causing any problems as far as I can tell, but still annoying.~~
  (this problem has been fixed in the latest commit)
- python setup.py test won't work anymore, I can make it work but this
  is not advised by pytest:
  https://github.com/pytest-dev/pytest-runner/#deprecation-notice
- run_test.py is still there but it's not really referenced anymore from
  anywhere and the options it accepts are completely different as it's options
  were based on nose. I would say it should be removed entirely but for now
  it is basically just a wrapper around pytest that basically does nothing.
- Removed references to test attributes as currently they are not being
  used anywhere anyway, I guess we can add them back if there is some
  use for them later.
- A lot of tests are still marked to skip when really they should be marked
  with xfail. This is also affecting the RDFT test manifests and result in
  reports saying tests are skipped when really we know they will fail and
  they are only skipped for this reason. But there is no change here from
  before, and pytest makes it easier to dynamically do expected failures.

Special thanks to Wes Turner for his advice and inputs on this process.

#### Other info

Fixes #1414
Closes #1268

The work in this PR is based on #1268 which was started by @white-gecko.

Coverage can be seen at https://coveralls.io/github/RDFLib/rdflib

- [latest master](https://coveralls.io/builds/43565943): 82.864%
- [latest iwana-20211018T2122-pytest](https://coveralls.io/builds/43723888): 83.38%

#### Proposed Changes                                                                                                                                                          
 - Replace nose with pytest
 - Use a pytest plugin for EARL reporting

nicholascar commented on .travis.yml:
```diff
@@ -36,8 +36,7 @@ before_script:
 script:
-    - PYTHONWARNINGS=default nosetests --with-timer --timer-top-n 42 --with-coverage --cover-tests --cover-package=rdflib
-    - coverage report
+    - pytest --cov
```
Travis is gone, we can remove all references to it

aucampia replied:
Removed now

nicholascar commented:
I've reviewing now! Please @ashleysommer and @white-gecko also review, this is a very important PR

white-gecko commented:
Thank you very much @aucampia for taking over this issue/pull-request.

I've just started my review.
I could not install the `berkeleydb` dev requirement, due to missing headers (this was introduced in #1347 and mentioned in https://github.com/RDFLib/rdflib/pull/1268#issuecomment-883789485 as an issue already. But we have to deal with this in a separate issue/pr.)
I have executed the tests and get a result of `84 warnings, 66 errors in 10.32s`. I assume the current setup is supposed to pass without any error as it does in drone.
All of the errors I get are `_pytest.pathlib.ImportPathMismatchError:`

e.g.

    ERROR build_build_ext/rdflib/collection.py - _pytest.pathlib.ImportPathMismatchError: ('rdflib.collection', '/home/natanael/…/rdflib/rdflib/collection.py', PosixPath('/home/natanael/…/rdflib/build_build_ext/rdflib/collection.py'))

So I don't know if this is good or bad.

aucampia commented:
> I have executed the tests and get a result of `84 warnings, 66 errors in 10.32s`. I assume the current setup is supposed to pass without any error as it does in drone. All of the errors I get are `_pytest.pathlib.ImportPathMismatchError:`

Can you share the full output log with https://gist.github.com/, and can you check if it passes if you run it with tox?

aucampia commented:
> I could not install the berkeleydb dev requirement, due to missing headers (this was introduced in #1347 and mentioned in #1268 (comment) as an issue already. But we have to deal with this in a separate issue/pr.)

I'm open to making another commit on this PR to fix this, just don't want to pile too many things on at once, but if this is blocking the evaluation of this then it may be worth expanding scope a bit.

aucampia commented:
looking now at warnings, I think maybe these can be eliminated by fixes/changes to `DefinedNamespace`:


```
test-runner_1           | ../usr/local/lib/python3.7/site-packages/_pytest/compat.py:312
test-runner_1           |   /usr/local/lib/python3.7/site-packages/_pytest/compat.py:312: UserWarning: Code: pytest_mock_example_attribute_that_shouldnt_exist is not defined in namespace BRICK
test-runner_1           |     return getattr(object, name, default)
test-runner_1           | 
test-runner_1           | ../usr/local/lib/python3.7/site-packages/_pytest/compat.py:312
test-runner_1           |   /usr/local/lib/python3.7/site-packages/_pytest/compat.py:312: UserWarning: Code: pytest_mock_example_attribute_that_shouldnt_exist is not defined in namespace DCAT
test-runner_1           |     return getattr(object, name, default)
test-runner_1           | 
test-runner_1           | ../usr/local/lib/python3.7/site-packages/_pytest/compat.py:312
test-runner_1           |   /usr/local/lib/python3.7/site-packages/_pytest/compat.py:312: UserWarning: Code: pytest_mock_example_attribute_that_shouldnt_exist is not defined in namespace PROF
test-runner_1           |     return getattr(object, name, default)
```

Will see what I can do, but will only have time to work on this on Friday.

nicholascar commented:
@white-gecko, did you try installing berkelyedb like this:

```
YES_I_HAVE_THE_RIGHT_TO_USE_THIS_BERKELEY_DB_VERSION=1 pip install berkeleydb
```

aucampia commented:
I added another commit to this PR e19fccea05e474174a5cd8a450fa49a1c6d0069b to ignore warnings with text

 - `pytest_mock_example_attribute_that_shouldnt_exist is not defined in namespace`
 - `_pytestfixturefunction is not defined in namespace`

These are a consequence of how pytest doctests detects mocks, and how pytest detects fixtures.

The other warnings that remain should be addressed but are not a consequence of switching to pytest.

aucampia commented:
@white-gecko @nicholascar  I added two more commits now


#### 4ea688796a983f533f6fdc345df06b1f33968de5

Added skipping of berkeleydb if it is not installed.

#### 41492440dccd2458c88cf008249dd9b3ef2db25c

Skipping was done incorrectly in test/test_issue200.py and failed on windows.

aucampia commented:
I'm trying to run the tests now on windows, but there are actually some problems that seem to be around from before. Some examples:

- rdflib.plugins.shared.jsonld.util.norm_url uses os.path to operate on URLs. This does not work on windows.
- mkstemp is not being used correctly, resulting in windows complaining that open files are being used.
- handling of file urIs in test/test_dawg.py won't work for windows because it simply drops the first 6 chars and this results in invalid windows paths.

I'm working now on another commit to fix all of these things, but I would be interested to know if all tests passed on windows for anyone.

white-gecko commented:
> @white-gecko, did you try installing berkelyedb like this:
> 
> ```
> YES_I_HAVE_THE_RIGHT_TO_USE_THIS_BERKELEY_DB_VERSION=1 pip install berkeleydb
> ```

No, that doesn't work. It is a message about a missing header file

```
FileNotFoundError: [Errno 2] No such file or directory: 'src/Modules/berkeleydb.h'
  Could not open module source to read the version number
```

The errors on my side do not appear to be related to the berkeleydb as they still occur on the latest commit.

> @white-gecko Can you share the full output log with https://gist.github.com/, and can you check if it passes if you run it with tox?
The full output is at: https://gist.github.com/white-gecko/43af9854c7c901b555a11fc29124552d

With `tox` I get more essentially the same output.

aucampia commented:
> The errors on my side do not appear to be related to the berkeleydb as they still occur on the latest commit.
> 
> > @white-gecko Can you share the full output log with https://gist.github.com/, and can you check if it passes if you run it with tox?
> The full output is at: https://gist.github.com/white-gecko/43af9854c7c901b555a11fc29124552d
> 
> With `tox` I get more essentially the same output.

@white-gecko can you clean your `__pycache__` and `.pyc` files and try again?

```bash
# IMPORTANT: Remove --dry-run flag

pipx run --spec pyclean pyclean --dry-run --verbose .

## OR

pip install --user --upgrade pyclean
pyclean --dry-run --verbose .
```

I will try setup a build for MacOS and Windows using GitHub actions and see if I can reproduce your problems, but like I said there seems to be some serious pre-existing issues on Windows which I'm still sorting out.

aucampia commented:
> The errors on my side do not appear to be related to the berkeleydb as they still occur on the latest commit.
> 
> > @white-gecko Can you share the full output log with https://gist.github.com/, and can you check if it passes if you run it with tox?
> The full output is at: https://gist.github.com/white-gecko/43af9854c7c901b555a11fc29124552d
> 
> With `tox` I get more essentially the same output.

@white-gecko you don't maybe have something related to Cython going on? The only thing I can find related to "build_build_ext" which I can find seems to be that: 

https://github.com/rchateauneu/rdflib/blob/dfdd4a336eaa3ed30b1d485009de7e2e426e1771/setup.py#L192-L202

white-gecko commented:
Cleaning the `__pycache__` and `.pyc` files didn't do the trick, but actually I had some Cython stuff left over from testing the Cython setup.

Now I get `3870 passed, 111 skipped, 7 xfailed, 1305 warnings in 35.48s`

aucampia commented:
> Now I get `3870 passed, 111 skipped, 7 xfailed, 1305 warnings in 35.48s`

Looks about right compared to drone results.

Still working on making tests pass on windows, to add another thing to the list of things broken on windows: the TSV result parser does not work on text files on windows as it does not handle `\r\n` line endings. I will probably just add xfail for all tsv result tests on windows, but I will look more into it tomorrow, but it seems these problems existed on master before, output of trying to run test_dawg.py from master on windows can be seen [here](https://gist.github.com/aucampia/6f4f2940a78f6c972291315454899b26) and it does not even get to the point of trying to parse tsv.

aucampia commented:
> Still working on making tests pass on windows

Just to be clear, I think this should be merged without fixes for windows, as those should really be treated as fixes in their own right given tests do not work on master under windows anyway, but working on it as at some point it would be nice to get some of the issues fixed, and the start to that is having tests that can actually run.

If there is still on this when I'm done with it I will add to this PR together with changes to run CI under windows and MacOS using GitHub actions.

nicholascar commented:
> I think this should be merged without fixes for windows

Agreed: we can address Windows issues later. 

I still can’t install BerkeleyDB on Windows (since I can’t compile 2015 C++) so there are a few things to fix.

Perhaps we need to add an option to just skill BerkeleyDB tests if BerkeleyDB isn’t installed, then at least all the other tests can be run on Windows.

Elsewhere I would like to get the BerkeleyDB tests running on Windows, but that’s more part of updating all the Stores, in particular the SQLite Store.

nicholascar commented:
On a Mac with the BerkeleyDB tests all running:

`7 failed, 3887 passed, 80 skipped, 7 xfailed, 1305 warnings, 7 errors in 209.89s`

The 7 errors are all due to _Can't assign requested address_, 5 of the 7 failures are also related to that and the remaining two are HTTPMock failures which I think are _also_ also related to that!

So all the failures & errors are all related to the same thing. What was the trick to allow it to assign an address? Running as sudo or something? Also, what else should I be doing to reduce the number of skipped tests?

On a Windows machine, without BerkeleyDB:

`669 failed, 3200 passed, 112 skipped, 7 xfailed, 1019 warnings in 167.46s`

All the test_dawg_data_sparql11 failed, just as @aucampia says, so just confirming that. The other failures seem to be Windows path related etc.

So, leaving aside the Windows results, my Mac results are close to yours.

aucampia commented:
> On a Mac with the BerkeleyDB tests all running:
> 
> `7 failed, 3887 passed, 80 skipped, 7 xfailed, 1305 warnings, 7 errors in 209.89s`
> 
> The 7 errors are all due to _Can't assign requested address_, 5 of the 7 failures are also related to that and the remaining two are HTTPMock failures which I think are _also_ also related to that!

@nicholascar please do share the output, I will see what I can do about it. I will try setup pipelines with GitHub actions on my branch shortly to see if that reproduces problems with MacOS.

nicholascar commented:
Hi @aucampia getting there. Just through I should do one more run with Fuseki available.

nicholascar commented:
Latest run results with Fuseki running to act as a read/write SPARQL endpoint: 

```
 11 failed, 3887 passed, 80 skipped, 7 xfailed, 1305 warnings, 7 errors in 194.20s 
```

Strange! So 4 more failures - auth errors, see the results details below - but the same number of passed and skipped  & errors. If there are 4 more failures, I expect to see 4 fewer passed or errors??

Here's my output:

```
FAILED test/test_dataset.py::t1::testDefaultGraph - urllib.error.HTTPError: HTTP Error 404: Not Found
FAILED test/test_dataset.py::t1::testGraphAware - urllib.error.HTTPError: HTTP Error 404: Not Found
FAILED test/test_dataset.py::t1::testIter - urllib.error.HTTPError: HTTP Error 404: Not Found
FAILED test/test_dataset.py::t1::testNotUnion - urllib.error.HTTPError: HTTP Error 404: Not Found
FAILED test/test_graph_http.py::TestGraphHTTP::test_3xx - OSError: [Errno 49] Can't assign requested address
FAILED test/test_graph_http.py::TestGraphHTTP::test_5xx - OSError: [Errno 49] Can't assign requested address
FAILED test/test_graph_http.py::TestGraphHTTP::test_content_negotiation - OSError: [Errno 49] Can't assign requested address
FAILED test/test_graph_http.py::TestGraphHTTP::test_source - OSError: [Errno 49] Can't assign requested address
FAILED test/test_sparqlstore.py::SPARQLMockTests::test_query - OSError: [Errno 49] Can't assign requested address
FAILED test/testutils.py::test.testutils.ServedSimpleHTTPMock
FAILED test/testutils.py::test.testutils.SimpleHTTPMock
ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_Query - OSError: [Errno 49] Can't assign requested address
ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_counting_graph_and_store_queries - OSError: [Errno 49] Can't assign requested address
ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_initNs - OSError: [Errno 49] Can't assign requested address
ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_noinitNs - OSError: [Errno 49] Can't assign requested address
ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_query_with_added_prolog - OSError: [Errno 49] Can't assign requested address
ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_query_with_added_rdf_prolog - OSError: [Errno 49] Can't assign requested address
ERROR test/test_sparqlupdatestore_mock.py::TestSPARQLConnector::test_graph_update - OSError: [Errno 49] Can't assign requested address
```

So it looks like all 11 of my errors (or 7, without Fuseki (!)) are a result of some OS port assignment thing and therefore "don't count" regarding evaluating this PR.

I'll try to reduce the numbers of skipped tests next

aucampia commented:
> ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_Query - OSError: [Errno 49] Can't assign requested address
> ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_counting_graph_and_store_queries - OSError: [Errno 49] Can't assign requested address
> ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_initNs - OSError: [Errno 49] Can't assign requested address
> ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_noinitNs - OSError: [Errno 49] Can't assign requested address
> ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_query_with_added_prolog - OSError: [Errno 49] Can't assign requested address
> ERROR test/test_sparqlstore.py::SPARQLStoreFakeDBPediaTestCase::test_query_with_added_rdf_prolog - OSError: [Errno 49] Can't assign requested address
> ERROR test/test_sparqlupdatestore_mock.py::TestSPARQLConnector::test_graph_update - OSError: [Errno 49] Can't assign requested address

I figured out what is wrong with these, macos does not like the random loopback IP addresses. I will add a commit to always use `127.0.0.1` shortly, just checking some other things.

nicholascar commented:
```
19 failed, 3818 passed, 67 skipped, 7 xfailed, 1297 warnings, 7 errors in 195.52s
```

Oh no! I installed `networkx`, fixed the Fuseki port allocation (3031, not 3030) and I've reduced the skipped count, but I've increased the failures and reduced the passed!

Here's my summary of skipped tests (using `pytest -rsx test`)

```
SKIPPED [50] test/test_dawg.py:334: Skipped
SKIPPED [2] test/test_dawg.py:205: Skipped
SKIPPED [1] test/test_extras_external_graph_libs.py:57: couldn't find graph_tool
SKIPPED [1] test/test_swap_n3.py:75: syntax_trailing_semicolon skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_too_nested skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_zero_objects skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_equals1 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_this_rules skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_this_quantifiers skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_thisadoc skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_keywords3 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_single_quote skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_formula_predicate skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_zero_predicates skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_literal_predicate skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_equals2 skipped, known issue
SKIPPED [1] test/test_trig.py:119: Iterative serialization currently produces 16 copies of everything
```

OK, clearly I'm able to get the skipped count down to 67 and, with different config, get 3887 tests passed - not quite 3901 but close!

I'll pass this now and leave final testing approval to @white-gecko (and perhaps @ashleysommer - please test!!).

nicholascar reviewed (approved):
Pretty close to everything working on my Mac

nicholascar commented:
> I figured out what is wrong with these, macos does not like the random loopback IP addresses.

Yes, that's right. I remembered that I've seen this before so, as I say, this is a Mac thing, not an RDFlib problem really!

> I will add a commit to always use 127.0.0.1 shortly, just checking some other things.

That would be cool

aucampia commented:
Added two more commits:

### c3270b18b1da20607db440bee531839d961e5746 - Change mock HTTP server to listen on 127.0.0.1 by default

This is so that it works properly on MacOS which does not permit
listening on random loopback addresses for user processes.

### 5393e0f34f42791576656aa3e702671a96cff53c - Fix how black errors are ingored in .drone.yaml

Errors were being piped to true, but that will fail if it runs with
pipefail. The better option is to do `black ... || true` which will work
for ignoring errors even with pipefail.

aucampia commented:
I setup a GitHub actions workflow to test on macos, results can be seen here: https://github.com/iafork/rdflib/actions?query=branch%3Aiwana-20211113T1312-github-actions

The branch is essentially the same as this one just with some changes for making github actions work:  https://github.com/iafork/rdflib/compare/iwana-20211018T2122-pytest...iwana-20211113T1312-github-actions

```
=== 3870 passed, 111 skipped, 7 xfailed, 1380 warnings in 101.80s (0:01:41) ====
```

Did not do much to reduce skipped tests, will maybe look at that later. May be worth enabling github actions on main repo, at the very least we can use it for additional validation but I think it is likely better than drone in most ways and runs macos, windows and linux.

aucampia commented:
> I setup a GitHub actions workflow to test on macos, results can be seen here: https://github.com/iafork/rdflib/actions?query=branch%3Aiwana-20211113T1312-github-actions
> 
> The branch is essentially the same as this one just with some changes for making github actions work: [iafork/rdflib@iwana-20211018T2122-pytest...iwana-20211113T1312-github-actions](https://github.com/iafork/rdflib/compare/iwana-20211018T2122-pytest...iwana-20211113T1312-github-actions)
> 
> ```
> === 3870 passed, 111 skipped, 7 xfailed, 1380 warnings in 101.80s (0:01:41) ====
> ```

Got it to [this](https://github.com/iafork/rdflib/actions/runs/1464457735) now:

```
=========================== short test summary info ============================
SKIPPED [4] test/test_dataset.py:37: Dependencies for store 'BerkeleyDB' not available!
SKIPPED [50] test/test_dawg.py:334: Skipped
SKIPPED [2] test/test_dawg.py:205: Skipped
SKIPPED [1] test/test_extras_external_graph_libs.py:57: couldn't find graph_tool
SKIPPED [10] test/test_graph.py:29: Dependencies for store 'BerkeleyDB' not available!
SKIPPED [10] test/test_graph_context.py:20: Dependencies for store 'BerkeleyDB' not available!
SKIPPED [1] test/test_graph_formula.py:27: Dependencies for store 'BerkeleyDB' not available!
SKIPPED [1] test/test_store_berkeleydb.py:89: skipping as berkleydb is missing
SKIPPED [1] test/test_store_berkeleydb.py:120: skipping as berkleydb is missing
SKIPPED [1] test/test_store_berkeleydb.py:55: skipping as berkleydb is missing
SKIPPED [1] test/test_store_berkeleydb.py:78: skipping as berkleydb is missing
SKIPPED [1] test/test_store_berkeleydb.py:64: skipping as berkleydb is missing
SKIPPED [1] test/test_store_berkeleydb.py:32: skipping as berkleydb is missing
SKIPPED [1] test/test_swap_n3.py:75: syntax_this_rules skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_too_nested skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_this_quantifiers skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_literal_predicate skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_keywords3 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_zero_predicates skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_equals2 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_equals1 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_trailing_semicolon skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_formula_predicate skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_thisadoc skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_zero_objects skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_single_quote skipped, known issue
SKIPPED [1] test/test_trig.py:119: Iterative serialization currently produces 16 copies of everything
XFAIL test/test_canonicalization.py::TestConsistency::test_consistent_ids
  reason: 
XFAIL test/test_diff.py::TestDiff::test_subsets
  reason: 
XFAIL test/test_issue190.py::test1
  Known issue with newlines in text
XFAIL test/test_issue190.py::test2
  Known issue with newlines in text
XFAIL test/test_literal.py::TestXsdLiterals::test_make_literals_ki
  reason: 
XFAIL test/test_namespace.py::ClosedNamespaceTest::test_repr_ef
  reason: 
XFAIL test/test_util.py::TestUtilTermConvert::test_util_from_n3_not_escapes_xf
  reason: 
==== 3887 passed, 98 skipped, 7 xfailed, 1372 warnings in 117.54s (0:01:57) ====
```

aucampia commented:
> > I setup a GitHub actions workflow to test on macos, results can be seen here: https://github.com/iafork/rdflib/actions?query=branch%3Aiwana-20211113T1312-github-actions
> > The branch is essentially the same as this one just with some changes for making github actions work: [iafork/rdflib@iwana-20211018T2122-pytest...iwana-20211113T1312-github-actions](https://github.com/iafork/rdflib/compare/iwana-20211018T2122-pytest...iwana-20211113T1312-github-actions)
> > ```
> > === 3870 passed, 111 skipped, 7 xfailed, 1380 warnings in 101.80s (0:01:41) ====
> > ```
> 
> Got it to [this](https://github.com/iafork/rdflib/actions/runs/1464457735) now:

Okay got it to pass even more on MacOS now: https://github.com/iafork/rdflib/runs/4218497302?check_suite_focus=true

```
=========================== short test summary info ============================
SKIPPED [50] test/test_dawg.py:334: Skipped
SKIPPED [2] test/test_dawg.py:205: Skipped
SKIPPED [1] test/test_extras_external_graph_libs.py:57: couldn't find graph_tool
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_literal_predicate skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_keywords3 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_thisadoc skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_zero_objects skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_equals2 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_zero_predicates skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_equals1 skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_too_nested skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_trailing_semicolon skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_formula_predicate skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_neg_single_quote skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_this_quantifiers skipped, known issue
SKIPPED [1] test/test_swap_n3.py:75: syntax_this_rules skipped, known issue
SKIPPED [1] test/test_trig.py:119: Iterative serialization currently produces 16 copies of everything
XFAIL test/test_canonicalization.py::TestConsistency::test_consistent_ids
  reason: 
XFAIL test/test_diff.py::TestDiff::test_subsets
  reason: 
XFAIL test/test_issue190.py::test1
  Known issue with newlines in text
XFAIL test/test_issue190.py::test2
  Known issue with newlines in text
XFAIL test/test_literal.py::TestXsdLiterals::test_make_literals_ki
  reason: 
XFAIL test/test_namespace.py::ClosedNamespaceTest::test_repr_ef
  reason: 
XFAIL test/test_util.py::TestUtilTermConvert::test_util_from_n3_not_escapes_xf
  reason: 
==== 3918 passed, 67 skipped, 7 xfailed, 1309 warnings in 142.71s (0:02:22) ====
```

Not sure things will get much better than this.

nicholascar commented:
Yes, I think that's going to be as good as it gets too!

@ashleysommer @white-gecko I've already indicated 'Approve'. Can one or other of you please do likewise and we will merge in. Then we will need to instantly raise an issue to address the DAWG tests.

aucampia commented:
> Then we will need to instantly raise an issue to address the DAWG tests.

I am almost done fixing issues on windows: https://github.com/iafork/rdflib/compare/iwana-20211018T2122-pytest...iwana-20211116T1039-fix_windows

I already have all DAWG tests passing there and just have like 50 failing tests left which I will try and fix soon.

nicholascar commented:
@aucampia your contributions to RDFlib recently have been truly inspiring!

white-gecko commented:
> @ashleysommer @white-gecko I've already indicated 'Approve'. Can one or other of you please do likewise and we will merge in.

I'm already in the middle of the review but not yet had the chance to finish it. I hope I will complete it soon.

white-gecko reviewed (approved):
This is a huge pull request. I could not check each individual line. Overall it looks good and it is working as expected. I'm very impressed by your work @aucampia, thank you for this huge contribution!

nicholascar commented:
Great work @aucampia indeed!

I will merge in tomorrow when I’m back at my computer (not phone like this!)

aucampia commented:
Glad to help out @white-gecko and @nicholascar - will hopefully have a PR for windows fixes this weekend and I will make an issue for running MacOS and windows pipelines so we can avoid issues in future where things break on one platform.
